src/azure-cli/azure/cli/command_modules/role/custom.py

```diff
@@ -280,8 +280,12 @@ def list_role_assignment_change_logs(cmd, start_time=None, end_time=None):  # py
     result = []
     worker = MultiAPIAdaptor(cmd.cli_ctx)
     start_events, end_events, offline_events, client = _get_assignment_events(cmd.cli_ctx, start_time, end_time)
-    role_defs = {d.id: [worker.get_role_property(d, 'role_name'),
-                        d.id.split('/')[-1]] for d in list_role_definitions(cmd)}
+
+    # Use the resource `name` of roleDefinitions as keys, instead of `id`, because `id` can be inherited.
+    #   name: b24988ac-6180-42a0-ab88-20f7382dd24c
+    #   id: /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c  # pylint: disable=line-too-long
+    role_defs = {d.name: [worker.get_role_property(d, 'role_name'), d.id.split('/')[-1]]
```

Comment:
Are the names returned by `list_role_definitions` always different ?

Reply:
Yes. It is a UUID and always unique.

Reply:
As tested, creating a `RoleDefinition` on sub-resource will always fallback to the subscription:

```
> az rest -m PUT -u https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleDefinitions/05675f70-84a5-44b5-94bf-2a6354b8077d?api-version=2018-01-01-preview -b "@role-definition-rest.json"
{
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/05675f70-84a5-44b5-94bf-2a6354b8077d",
  "name": "05675f70-84a5-44b5-94bf-2a6354b8077d",
...
```

<details><summary> role-definition-rest.json </summary>

```
{
    "properties": {
        "roleName": "Contoso On-call 4",
        "type": "CustomRole",
        "description": "Perform VM actions and read storage and network information.",
        "assignableScopes": [
            "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1"
        ],
        "permissions": [
            {
                "actions": [
                    "Microsoft.Compute/*/read",
                    "Microsoft.Compute/virtualMachines/start/action",
                    "Microsoft.Compute/virtualMachines/restart/action",
                    "Microsoft.Network/*/read",
                    "Microsoft.Storage/*/read",
                    "Microsoft.Authorization/*/read",
                    "Microsoft.Resources/subscriptions/resourceGroups/read",
                    "Microsoft.Resources/subscriptions/resourceGroups/resources/read",
                    "Microsoft.Insights/alertRules/*",
                    "Microsoft.Support/*"
                ],
                "notActions": [],
                "dataActions": [
                    "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/*"
                ],
                "notDataActions": [
                    "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write"
                ]
            }
        ]
    },
    "type": "Microsoft.Authorization/roleDefinitions"
}
```

</details>

Notice that `id` doesn't have `/resourceGroups/rg1`.